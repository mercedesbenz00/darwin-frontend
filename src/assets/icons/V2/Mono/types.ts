export enum IconVariant {
  DEFAULT = 'default',
  DISABLED = 'disabled',
  SELECTED = 'selected',
  NEGATIVE = 'negative',
}
