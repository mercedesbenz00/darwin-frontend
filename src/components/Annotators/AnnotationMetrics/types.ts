import { MembershipPayload, DatasetAnnotationReportPayload } from '@/store/types'

export type AnnotationDataRange = 'total' | 'month' | 'week' | 'day'
export type AnnotationDataGranularity = 'day' | 'hour'

export type ChartDimension =
  'annotationTime' |
  'imagesAnnotated' |
  'avgTimePerAnnotation' |
  'annotationsCreated' |
  'reviewPassRate'

export type AdaptedDataPoint =
  {[k in Exclude<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number } &
  {[k in Extract<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number | null} &
  { userId: number | null, date: string }

export type AggregatedDataPoint = Pick<
    DatasetAnnotationReportPayload,
    | 'annotation_time'
    | 'annotations_approved'
    | 'annotations_created'
    | 'date'
    | 'images_annotated'
    | 'images_approved'
    | 'images_rejected'
    | 'user_id'
  >

export type AnnotationDimensionData = {
  annotationTime: number,
  annotationsCreated: number,
  avgTimePerAnnotation: number | null,
  imagesAnnotated: number,
  reviewPassRate: number | null
}

export type AnnotationChartData = {
  data: AdaptedDataPoint[]
  dimensions: AnnotationDimensionData
  member?: MembershipPayload
  visible: boolean
}

export type AnnotationBarChartData = {
  data: AggregatedDataPoint[]
  member?: MembershipPayload
}

/**
 * Format for a single item to be shown in a chart tooltip.
 *
 * A tooltip will contain multiple items
 */
export type ChartTooltipData = { member: MembershipPayload | null, value: string}
