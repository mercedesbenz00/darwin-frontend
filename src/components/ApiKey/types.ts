export type AbilityGroup =
  'manage_comments' |
  'manage_tasks' |
  'manage_datasets' |
  'manage_team' |
  'manage_team_members' |
  'manage_models'

export type Group = { id: AbilityGroup; name: string }
export type GroupOption = Group & { selected: boolean }

export enum SupportedLanguages {
  shell = 'shell',
  python = 'python',
  javascript = 'javascript',
  elixir = 'elixir',
  cli = 'cli'
}

export type AccessCodeExample = { language: SupportedLanguages, code: string }
