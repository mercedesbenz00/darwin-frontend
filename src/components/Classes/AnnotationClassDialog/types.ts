import { AnnotationClassPayload, AnnotationClassMetadata, AnnotationTypeName } from '@/store/types'

type PendingFields =
  | 'crop_key'
  | 'crop_url'
  | 'image_height'
  | 'image_width'
  | 'key'
  | 'original_image_url'
  | 'scale'
  | 'x'
  | 'y'

/**
 * Holds a class image that isn't necessarily persisted to the backend yet.
 * That means it might be missing any of the url fields
 *
 * It can be saved to backend once at least the `key` field is set, as that one
 * can be used to infer/compute all the other fields.
 *
 * - `crop_key` - repleace "original" with "crop" in `key` field
 * - `crop_url` - signed s3 url (by backend), for `crop_key`
 * - `original_image_url` - signed s3 url (by backend) for `key`
 */
export type PendingClassImage =
  Omit<AnnotationClassPayload['images'][0], PendingFields>
  & Partial<Pick<AnnotationClassPayload['images'][0], PendingFields>>

// eslint-disable-next-line camelcase
export type PendingClassImageWithUrlSet = PendingClassImage & { original_image_url: string }

export const isClassImage = (
  image: PendingClassImage
): image is AnnotationClassPayload['images'][0] =>
  !!image.key && !!image.crop_key

export type AnnotationClassData = {
  datasets: AnnotationClassPayload['datasets']
  description: string | null
  hotkey: null | string
  id: number | null
  images: PendingClassImage[]
  metadata: AnnotationClassMetadata
  name: string | null
}

export type DefaultAnnotationClassData = Partial<AnnotationClassData> & {
  selectedMainAnnotationType?: AnnotationTypeName
}

export type AnnotationClassValidationErrors = {
  annotationTypes?: string
  name?: string
  skeleton?: string
}

export const VIEWPORT_WIDTH: number = 150
export const VIEWPORT_HEIGHT: number = 150

export type CropInfo = {
  orientation: number
  points: [string, string, string, string]
  zoom: number
}

/**
 * VueCroppie is really archaic and requires us to get crop info via $ref,
 * so this type defines that ref
 */
export type Croppie = {
  /** Binds options to the underlying croppie library */
  bind: (options: Pick<CropInfo, 'points' | 'zoom'> & { url: string}) => void

  /** Retrieves current crop info */
  get: () => CropInfo
  result: (options?: Record<string, string>, callback?: Function) => void

  /** Sets scale/zoom level of cropper */
  setZoom: (value: number) => void
}
