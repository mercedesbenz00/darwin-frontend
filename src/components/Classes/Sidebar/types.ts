import { AnnotationTypeName } from '@/store/types'

export type AnnotationTypeFilterItemType = {
  id: string | number
  label: string
  icon?: string
  name: AnnotationTypeName
  color: string
  count: number | null
  [key: string]: any
}
