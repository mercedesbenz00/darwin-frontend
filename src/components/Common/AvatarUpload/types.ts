export type AvatarUploadData = {
  hash: string
  file: File
  type: string | null
}
