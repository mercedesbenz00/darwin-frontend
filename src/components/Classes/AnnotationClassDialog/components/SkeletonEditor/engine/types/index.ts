export * from './SkeletonEdgeType'
export * from './SkeletonNodeType'
