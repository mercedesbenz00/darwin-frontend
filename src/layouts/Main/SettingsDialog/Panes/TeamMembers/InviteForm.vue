<template>
  <div
    v-if="canInvite"
    v-tooltip="partnerTooltip"
    class="members__new"
  >
    <div class="members__new__input">
      <input-field
        id="new-invitation"
        ref="emailInput"
        v-model="newInvitationEmail"
        :disabled="isPartnerMember"
        type="text"
        placeholder="+ Enter email and press enter to invite"
        theme="light"
        @enter="createInvitation"
      />
    </div>
    <div class="members__new__role">
      <role-dropdown
        id="role"
        ref="roleDropdown"
        v-model="newInvitationRole"
        name="role"
        placeholder="Role"
        :options="invitationRoleOptions"
        :disabled="isPartnerMember"
      />
    </div>
    <button
      class="members__new__send"
      :class="{'members__new__send--disabled': !newInvitationEmail}"
      :disabled="isPartnerMember"
      @click="createInvitation"
    >
      <img src="/static/imgs/send.svg">
    </button>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, nextTick, onMounted, ref } from 'vue'

import InputField from '@/components/Common/InputField/V1/InputField.vue'
import RoleDropdown from '@/components/Common/RoleDropdown.vue'
import { useAuth } from '@/composables'
import { useStore } from '@/composables/useStore'
import { useTeamStore } from '@/composables/useTeamStore'
import { MembershipRole } from '@/store/types'

import { invitationRoleOptions as getRoleOptions, ROLE_LABELS } from './data'

type InvitationError = {
  email?: string
  role?: string
}

export default defineComponent({
  name: 'InviteForm',
  components: { InputField, RoleDropdown },
  setup () {
    const store = useStore()

    const emailInput = ref<InputField | null>()

    const teamStore = useTeamStore()

    const { isAuthorized } = useAuth()
    const invitationRoleOptions = computed(() => getRoleOptions(isAuthorized))

    const canInvite = computed<boolean>(() => invitationRoleOptions.value.length > 0)

    const user = computed(() => store.state.user.profile)

    const isPartnerMember = computed<boolean>(() => {
      const currentTeam = teamStore.currentTeam.value
      if (!currentTeam) { return false }

      const currentUser = user.value
      if (!currentUser) { return false }

      // current team is not client, so the whole concept doesn't apply
      if (!(currentTeam.managed_status === 'client' && !!currentTeam.partner_id)) {
        return false
      }

      // current team is client and user it direct member of client
      if (teamStore.currentTeamMemberships.value.some(m => m.user_id === currentUser.id)) {
        return false
      }

      // current team is client and user is direct member of partner
      return teamStore.allRelevantMemberships.value.some(
        m => m.team_id === currentTeam.partner_id && m.user_id === currentUser.id
      )
    })

    const partnerTooltip = computed<string | undefined>(() => {
      if(!isPartnerMember.value) { return }
      return 'You are not allowed to add users directly to this team'
    })

    const newInvitationRole = ref<MembershipRole>('member')
    const newInvitationEmail = ref<string>('')

    const roleLabels = ROLE_LABELS

    onMounted(() => {
      if (invitationRoleOptions.value.length > 0) {
        newInvitationRole.value = invitationRoleOptions.value[0].id
      }
    })

    const error = ref<InvitationError>({})

    const updateErrors = (): void => {
      nextTick(() => {
        if (!error.value.email) { return }
        emailInput.value?.setError(error.value.email)
      })
    }

    const createInvitation = async (): Promise<void> => {
      if (newInvitationEmail.value.length === 0) { return }
      const validationErrors =
        await teamStore.createInvitation(newInvitationEmail.value, newInvitationRole.value)

      if (validationErrors) {
        error.value = validationErrors
        updateErrors()
        return
      }

      newInvitationEmail.value = ''
    }

    return {
      canInvite,
      createInvitation,
      emailInput,
      error,
      invitationRoleOptions,
      isPartnerMember,
      newInvitationEmail,
      newInvitationRole,
      partnerTooltip,
      roleLabels
    }
  }
})
</script>

<style lang="scss" scoped>
.members__new {
  @include row--center;
  margin: 10px 84px 30px 12px;
}

.members__new__input {
  flex: 1;
  margin-right: 15px;
}

.members__new__role {
  @include row--center;
  width: 180px;
  margin-right: 30px;
}

.members__new__send {
  @include row--center;
  width: 35px;
  cursor: pointer;
  background: transparent;

  &--disabled {
    opacity: .3;
  }

  &:not(.members__new__send--disabled) {
    &:hover {
      opacity: .7;
    }

    &:active {
      opacity: .5;
    }
  }
}
</style>
