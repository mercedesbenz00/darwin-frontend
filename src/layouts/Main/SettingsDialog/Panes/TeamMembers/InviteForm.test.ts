import { createLocalVue, shallowMount, Wrapper } from '@vue/test-utils'
import Vuex from 'vuex'

import createTestStore from 'test/unit/createTestStore'
import { stubDirectiveWithAttribute } from 'test/unit/directiveStubs'
import {
  buildMembershipPayload,
  buildTeamPayload,
  buildUserPayload
} from 'test/unit/factories'

import InviteForm from '@/layouts/Main/SettingsDialog/Panes/TeamMembers/InviteForm.vue'
import { Ability, MembershipRole } from '@/store/types'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.directive('tooltip', stubDirectiveWithAttribute)

let store: ReturnType<typeof createTestStore>

const ability: Ability = { actions: ['manage_invitations'], subject: 'all' }

beforeEach(() => {
  store = createTestStore()
  store.commit('auth/SET_ABILITIES', [ability])
})

class Model {
  wrapper: Wrapper<Vue>

  constructor (wrapper: Wrapper<Vue>) {
    this.wrapper = wrapper
  }

  get emailField (): Wrapper<Vue> {
    return this.wrapper.find('input-field-stub')
  }

  get email (): { wrapper: Wrapper<Vue>, disabled: boolean } {
    const wrapper = this.emailField
    return {
      wrapper,
      disabled: wrapper.props('disabled')
    }
  }

  setEmail (email: string): void {
    this.emailField.vm.$emit('input', email)
  }

  get roleField (): Wrapper<Vue> {
    return this.wrapper.find('role-dropdown-stub')
  }

  get role (): { wrapper: Wrapper<Vue>, disabled: boolean, value: string } {
    const wrapper = this.roleField
    return {
      wrapper,
      disabled: wrapper.props('disabled'),
      value: wrapper.props('value')
    }
  }

  setRole (role: MembershipRole): void {
    this.roleField.vm.$emit('input', role)
  }

  get button (): { wrapper: Wrapper<Vue>, disabled: boolean } {
    const wrapper = this.wrapper.find('button')
    return {
      wrapper,
      disabled: wrapper.attributes('disabled') === 'disabled'
    }
  }

  clickButton (): Promise<void> | void {
    return this.button.wrapper.trigger('click')
  }
}

const itMatchesSnapshot = (): void => it('matches snapshot', () => {
  const wrapper = shallowMount(InviteForm, { localVue, store })
  expect(wrapper).toMatchSnapshot()
})

describe('when client team', () => {
  const partner = buildTeamPayload({ id: 1, managed_status: 'partner' })
  const client = buildTeamPayload({
    id: 2, managed_status: 'client', partner, partner_id: partner.id
  })

  beforeEach(() => {
    store.commit('team/SET_CURRENT_TEAM', client)
  })

  describe('viewed as partner team member', () => {
    beforeEach(() => {
      const user = buildUserPayload({ id: 9 })
      store.commit('user/SET_PROFILE', user)
      store.commit('team/SET_MEMBERSHIPS', [
        buildMembershipPayload({ id: 1, user_id: user.id, team_id: partner.id })
      ])
    })

    itMatchesSnapshot()

    it('renders as disabled', () => {
      const wrapper = shallowMount(InviteForm, { localVue, store })
      const model = new Model(wrapper)
      expect(model.email.disabled).toBe(true)
      expect(model.role.disabled).toBe(true)
      expect(model.button.disabled).toBe(true)
    })
  })
})

describe('when regular team', () => {
  const team = buildTeamPayload({ id: 90, managed_status: 'regular' })
  const user = buildUserPayload({ id: 9 })
  beforeEach(() => {
    store.commit('team/SET_CURRENT_TEAM', team)
  })

  describe('as full member', () => {
    beforeEach(() => {
      store.commit('user/SET_PROFILE', user)
      store.commit('team/SET_MEMBERSHIPS', [
        buildMembershipPayload({ id: 1, user_id: user.id, team_id: team.id })
      ])
    })

    itMatchesSnapshot()

    it('creates new invitation', async () => {
      const wrapper = shallowMount(InviteForm, { localVue, store })

      const email = 'another_invitation@example.com'
      const role = 'member'

      const model = new Model(wrapper)

      await model.setEmail(email)
      await model.setRole(role)
      await model.clickButton()

      expect(store.dispatch).toHaveBeenCalledWith('team/addInvitations', {
        teamId: team.id,
        invitations: [{ email, role }]
      })
    })
  })

  describe('as workforce manager', () => {
    const ability: Ability = {
      actions: ['manage_invitations'], subject: 'invitation', conditions: { role: 'annotator' }
    }
    beforeEach(() => {
      const membership = buildMembershipPayload({
        id: 1,
        user_id: user.id,
        team_id: team.id,
        role: 'workforce_manager'
      })

      store.commit('team/SET_MEMBERSHIPS', [membership])
      store.commit('auth/SET_ABILITIES', [ability])
    })

    itMatchesSnapshot()

    test('preselects worker role', async () => {
      const wrapper = shallowMount(InviteForm, { localVue, store })
      await wrapper.vm.$nextTick()

      const model = new Model(wrapper)

      expect(model.role.value).toEqual('annotator')
    })
  })
})
