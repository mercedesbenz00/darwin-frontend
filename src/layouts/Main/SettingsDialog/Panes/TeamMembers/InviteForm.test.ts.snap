// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when client team viewed as partner team member matches snapshot 1`] = `
<div class="members__new" tooltip="You are not allowed to add users directly to this team">
  <div class="members__new__input">
    <input-field-stub value="" id="new-invitation" placeholder="+ Enter email and press enter to invite" type="text" disabled="true" theme="light"></input-field-stub>
  </div>
  <div class="members__new__role">
    <role-dropdown-stub id="role" value="member" name="role" options="[object Object],[object Object],[object Object],[object Object]" placeholder="Role" disabled="true"></role-dropdown-stub>
  </div> <button disabled="disabled" class="members__new__send members__new__send--disabled"><img src="/static/imgs/send.svg"></button>
</div>
`;

exports[`when regular team as full member matches snapshot 1`] = `
<div class="members__new" tooltip="undefined">
  <div class="members__new__input">
    <input-field-stub value="" id="new-invitation" placeholder="+ Enter email and press enter to invite" type="text" theme="light"></input-field-stub>
  </div>
  <div class="members__new__role">
    <role-dropdown-stub id="role" value="member" name="role" options="[object Object],[object Object],[object Object],[object Object]" placeholder="Role"></role-dropdown-stub>
  </div> <button class="members__new__send members__new__send--disabled"><img src="/static/imgs/send.svg"></button>
</div>
`;

exports[`when regular team as workforce manager matches snapshot 1`] = `
<div class="members__new" tooltip="undefined">
  <div class="members__new__input">
    <input-field-stub value="" id="new-invitation" placeholder="+ Enter email and press enter to invite" type="text" theme="light"></input-field-stub>
  </div>
  <div class="members__new__role">
    <role-dropdown-stub id="role" value="member" name="role" options="[object Object]" placeholder="Role"></role-dropdown-stub>
  </div> <button class="members__new__send members__new__send--disabled"><img src="/static/imgs/send.svg"></button>
</div>
`;
